p2p/net/conn.go

```diff
@@ -143,111 +163,164 @@ func (c *FormattedConnection) publish(message []byte) {
 	c.networker.EnqueueMessage(IncomingMessageEvent{c, message})
 }
 
-// incomingChannel returns the incoming messages channel
-func (c *FormattedConnection) incomingChannel() chan []byte {
-	return c.formatter.In()
-}
+// NOTE: this is left here intended to help debugging in the future.
+//func (c *FormattedConnection) measureSend() context.CancelFunc {
+//	ctx, cancel := context.WithCancel(context.Background())
+//	go func(ctx context.Context) {
+//		timer := time.NewTimer(time.Second * 20)
+//		select {
+//		case <-timer.C:
+//			i := crypto.UUIDString()
+//			c.logger.With().Info("sending message is taking more than 20 seconds", log.String("peer", c.RemotePublicKey().String()), log.String("file", fmt.Sprintf("/tmp/stacktrace%v", i)))
+//			buf := make([]byte, 1024)
+//			for {
+//				n := runtime.Stack(buf, true)
+//				if n < len(buf) {
+//					break
+//				}
+//				buf = make([]byte, 2*len(buf))
+//			}
+//			err := ioutil.WriteFile(fmt.Sprintf("/tmp/stacktrace%v", i), buf, 0644)
+//			if err != nil {
+//				c.logger.Error("ERR WIRTING FILE %v", err)
+//			}
+//		case <-ctx.Done():
+//			return
+//		}
+//	}(ctx)
+//	return cancel
+//}
 
-// Send binary data to a connection
-// data is copied over so caller can get rid of the data
-// Concurrency: can be called from any go routine
 func (c *FormattedConnection) Send(m []byte) error {
-	err := c.formatter.Out(m)
+	c.wmtx.Lock()
+	defer c.wmtx.Unlock()
+	if c.closed {
+		return fmt.Errorf("connection was closed")
+	}
+
+	c.deadliner.SetWriteDeadline(time.Now().Add(c.deadline))
+	_, err := c.w.WriteRecord(m)
 	if err != nil {
+		cerr := c.closeUnlocked()
+		if cerr != ErrAlreadyClosed {
+			c.networker.publishClosingConnection(ConnectionWithErr{c, err}) // todo: reconsider
+		}
+		return err
+	}
+	//metrics.PeerRecv.With(metrics.PeerIdLabel, c.remotePub.String()).Add(float64(len(m)))
```

Comment:
remove this?

Reply:
I remove the comment and bring back the line since I was just testing temporarily without using metrics